src/base/porting.h

```diff
@@ -317,6 +317,10 @@ extern int free_space (const char *, int);
 #define snprintf_dots_truncate(dest, max_len, ...) \
   if (snprintf (dest, max_len, __VA_ARGS__) < 0) \
     snprintf (dest + max_len - 4, 4, "...")
+#define strncpy_size(buf, str, size) \
+  strncpy (buf, str, size); buf[(size) - 1] = '\0'
+#define strncpy_bufsize(buf, str) \
+  strncpy_size (buf, str, sizeof (buf))
```

Comment:
I propose to add a static assert for buf to be an array.
```suggestion
  check_is_array (buf); strncpy_size (buf, str, sizeof (buf))
```

>#include <iostream>
>#include <type_traits>
>#include <typeinfo>
>
>template <class T>
>void check_is_array ( const T &a )
>{
>  static_assert (std::is_array<T>::value == 1, "expect array");
>}
>int main()
>{
> float a[10];
> int *b;
> check_is_array (a);   // ok
> check_is_array (b);   // error
>}